Multi target grad

hotpxl opened the pull request.

lryta commented on examples/basics/logistic.py:
```diff
@@ -3,7 +3,7 @@
 import minpy.dispatch.policy as policy
 
-#np.set_policy(policy.OnlyNumPyPolicy())
+np.set_policy(policy.OnlyNumPyPolicy())
```
Cannot use MXNet OPs?

hotpxl replied:
Sorry. I don't have GPU MXNet on my box. I'll revert this change.
Normally we shouldn't commit changes like this (changing file-wise policy or file-wise logging options).
My bad.

hotpxl commented:
@jermainewang 
Pls take a look. Will merge soon to avoid further conflicts and for @lryta to merge his additions. Will work on the following files to further clean up and change doc style:
* array.py
* core.py
* primitive.py

since they are closely related to this merge.

PS. There are lots of lint warnings in existing code. Travis check is lenient. Probably it's not needed for a mandatory check, but I expect to be more strict with ourselves.

jermainewang commented on VERSION:
```diff
@@ -1 +1 @@
-0.1.6
\ No newline at end of file
+0.1.6
```
We could bump version for this.

hotpxl replied:
Will bump on master after merge

jermainewang commented on minpy/core.py:
```diff
@@ -5,47 +5,57 @@
         arrays = tuple(Value.wrap(a) for a in args)
         argnums = [argnum] if isinstance(argnum, int) else argnum
         for i in argnums:
             arrays[i]._marked_for_bp = True
```
Delete? I think you've removed this field.

hotpxl replied:
Still needed to calculate `need_bp`

jermainewang commented on minpy/core.py:
```diff
@@ -5,47 +5,57 @@
         for i in argnums:
-            grad_vals.append(arrays[i].node.partial_derivative(
-                result_array.node))
             arrays[i]._marked_for_bp = False
```
Delete?

hotpxl replied:
This file is updated. refresh?

jermainewang commented on minpy/core.py:
```diff
@@ -125,14 +138,17 @@ def func(**kwargs):
                     executor.backward(out_grads=g)
                     ret = executor.grad_arrays[arg_names.index(keyname)]
                     return ret
+
```
Question: are we still using per keyword gradient definition? I thought we will have a way to define one gradient for a subset of keywords (or just simply all the keywords).

hotpxl replied:
Per keyword is still supported. Per index supports multiple gradient.

jermainewang commented on minpy/primitive.py:
```diff
@@ -94,57 +128,62 @@ def get_val(x, mutate):
+                        'Adding partial derivative to func "{}" on argument {}.'.
+                        format(self._func, i))
+                    grad_func_rec = self._grad_func[i]
+                    grad_func = grad_func_rec.f(result_value, *arg_values,
```
So the `f` field in `GradFunc` is like a gradient function selector? I would use another name.

hotpxl replied:
Did you pull the latest changes? Looks different on my side

jermainewang commented on minpy/primitive.py:
```diff
@@ -153,7 +192,7 @@ def def_grad(self, func, argnum=0):
+        self._grad_func[argnum] = GradFunc(f=func, multi_grad_indices=None)
         return self
 
     def def_grad_kw(self, func, key):
```
My idea is to redirect `def_grad` and `def_grad_kw` all to `def_multi_grad`. They are just special cases.

hotpxl replied:
Implemented multigrad upon single grad. Single grad is hashable and results in faster lookup. Multigrad is a special case.

jermainewang commented on minpy/primitive.py:
```diff
@@ -163,16 +202,46 @@ def def_grad_kw(self, func, key):
+
+        for argnum in argnums:
+            self._grad_func[argnum] = GradFunc(
+                f=func, multi_grad_indices=argnums)
```
Hmm... I'm not sure about this. I thought we will only have a map in primitive like this:
```
arg_index_set + keyword_set => gradient function on them
```

jermainewang commented:
Comments left. Haven't gone through the tape.py file yet. Will update soon.

lryta commented on minpy/tape.py:
```diff
@@ -0,0 +1,209 @@
+"""Tape for recording gradient calculation."""
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import with_statement
```
It seems that this is mandatory on after Python 2.6.

lryta commented on minpy/tape.py:
```diff
@@ -0,0 +1,209 @@
+            Primitive type indicating location of execution.
+        """
+
+        def add_owner(single_owner):
```
Does the argument name single_owner sounds a little ambiguous? I was wondering if the "if" logic of determining the type of "single_owner" could be moved into its parent function?

lryta replied:
Ok I see. Maybe just change the name of single_owner.

hotpxl replied:
It's an internal scoped function. Hard to come up with names. single_owner just describes the behavior when it recurses.

lryta commented on minpy/primitive.py:
```diff
@@ -94,57 +128,62 @@ def get_val(x, mutate):
         # Wrap the result raw value with wrapper and node.
-        result = Value.wrap(result_value, marked=need_bp)
+        result = array.Value.wrap(result_value, marked=need_bp)
         if need_bp:
```
Do we need an error handling here to check if tape.global_tape is None?

hotpxl replied:
tape.global_tape is assumed to not be None here. otherwise it's a fatal error

hotpxl commented:
@jermainewang I fixed the issues we discussed about the other day. Take a look at the last few commits.

For `cnn.py`, I first tried running using the old gradient code, but hacked executor to run only once (added a variable to ensure one ff matches one bp). I can get to around 60% acc. You told me it was around 70%. But 60% is what I got.
After I changed to multi-grad, I'm also getting 60%. Results on my side are consistent.

For `cnn_sym.py`, I'm getting a bunch of NaNs. I tried running code before my change, NaN as well. So this is consistent (but definitely not correct).

`mlp.py` converges around 50%.

jermainewang commented:
I see. How about the RL example and other unit tests?

hotpxl commented:
Unittests are fine.
I haven't tried RL example yet.

jermainewang commented:
Hi Yutian, just uploaded a commit (1f5c2afc0a5a31208eb37f9bee24c2ce9268cec1) to fix the NaN problem. The reason is the `SoftmaxOutput` is not correctly called. I revert it to our `softmax_loss`. Please have a try.

hotpxl commented:
@jermainewang I tried the changes and now `cnn_sym.py` converges to around 60%.
